### PR TITLE
Error when a solid/op is invoked within another solid/op

### PR DESCRIPTION
**Pull request by dpeng817**
Relevant issue: https://github.com/dagster-io/dagster/issues/5542

Right now, we kind of let people blindly invoke ops/solids within other ops/solids. This leads to misunderstandings with our composition DSL, and hinders ramp-up with dagster. 

This PR hard-errors when performing this type of invocation, and points people to the relevant documentation. Technically, this is a breaking change, but it's a breaking change for undefined behavior.

### CODE REVIEW

**Comment by schrockn** [comment]
We should have a method for direct invocation if we make this change. e.g. `some_op.invoke(context, arg1)`

**Comment by dpeng817** [comment]
> We should have a method for direct invocation if we make this change. e.g. some_op.invoke(context, arg1)

@schrockn do you mean in the general case, moving direct invocation to the `.invoke(...)` method? Or making `.invoke(...)` have the same behavior as direct invocation so that folks who for some reason want to directly invoke within their ops can do so

**Comment by schrockn** [comment]
"Or making .invoke(...) have the same behavior as direct invocation so that folks who for some reason want to directly invoke within their ops can do so"

^--- This exactly. Have the *ability* to do direct invocation in the cases where this PR is making it a hard error.

**Comment by alangenfeld** [comment]
I remember discussing this in person at some point and I think we had an argument for why hard error wasn't the way to go. Wish I could remember the details. 

Does anyone have strong opinions on error vs warn here? I soft lean towards warn instead of error - assuming the warning would actually show up somewhere useful. I guess it would have to fire an event log warning to be visible.

**Comment by dpeng817** [comment]
cc @alangenfeld @schrockn My take is that invocation within ops is a quite advanced / niche use case. Invoking an op within another op is using dagster in a way that wasn't intended, which means one of two things to me:
1. You are very familiar with Dagster, understand the graph DSL and are stretching the abstractions to suit your needs. For these folks, `.invoke()` should work fine.
2. You don't understand Dagster, and think you're building out a DAG by invoking ops within other ops. For these folks, probably don't want to be invoking ops within other ops and we should turn these folks away from that path. 
TL;dr, I think it's a workflow that is fine to support, but we should make sure that users are actively aware of the behavior that they are opting into, which is what makes a `.invoke()` method attractive.

File: python_modules/dagster/dagster_tests/core_tests/test_solid_invocation.py [hunk 1]
```diff
@@ -453,6 +453,10 @@ def solid_multiple_outputs():
 
 
 def test_wrong_output():
+    from dagster.core.definitions.composition import (  # pylint: disable=unused-import
```
Comment (dpeng817) [inline python_modules/dagster/dagster_tests/core_tests/test_solid_invocation.py:456]: I have no idea why this fixes test failures

File: python_modules/dagster/dagster/core/definitions/solid_definition.py [hunk 2]
```diff
@@ -137,61 +138,80 @@ def __init__(
...
+            enter_op_execution_context(self.name, f"@{self.node_type_str}")
+            try:
+                result = self.invoke(*args, **kwargs)
+            finally:
+                if is_in_op_execution_context():
+                    exit_op_execution_context()
+
+            return result
+
+    def invoke(self, *args, **kwargs) -> Any:
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/solid_definition.py:166]: doc block

File: python_modules/dagster/dagster/core/definitions/solid_definition.py [hunk 3]
```diff
@@ -137,61 +138,80 @@ def __init__(
...
+            exit_op_execution_context,
+        )
 
         if is_in_composition():
             return super(SolidDefinition, self).__call__(*args, **kwargs)
+        elif is_in_op_execution_context():
+            name, source = current_op_execution_context()
+            raise DagsterInvalidInvocationError(
+                f"Attempted to invoke @{self.node_type_str} '{self.name}' within {source} '{name}', which is undefined behavior. In order to compose invocations, check out the graph API: https://docs.dagster.io/concepts/ops-jobs-graphs/nesting-graphs#nesting-graphs"
+            )
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/solid_definition.py:155]: This should probably mention `invoke` method explicitly or how else would one discover it? Maybe update the linked docs as well with a section 

A little worried about this deep link breaking over time - not sure best way to defend against that.

**Review by alangenfeld — CHANGES_REQUESTED** [review]
alright looks like hard error is the way to go - lets just ensure `invoke` is actually discoverable

**Comment by dpeng817** [comment]
Closing for now until I have time to revisit
